source/mir/random/ndvariable.d

```diff
@@ -0,0 +1,293 @@
+    ///
+    enum isRandomVariable = true;
+
+    import mir.ndslice.sorting;
+    private size_t dim;
+
+    /++
+    Params:
+        dim = dimension of the simplex
+    +/
+    this(size_t dim)
+    {
+        this.dim = dim;
+    }
+
+    ///
+    void opCall(G)(ref G gen, T[] result)
+    {
+        assert(result.length == dim + 1);
+        opCall(gen, result.sliced);
+    }
+
+    ///
+    void opCall(G, SliceKind kind, Iterator)(ref G gen, Slice!(kind, [1], Iterator) result)
+        if (isSaturatedRandomEngine!G)
+    {
+        assert(result.length == dim + 1);
+
+        for(size_t i = 0; i < dim; ++i)
+            result[i] = gen.rand!T.fabs;
+        result[dim] = T(1);
+
+        sort(result[]);
+        for(size_t i = dim; i > 0; --i)
+            result[i] = result[i] - result[i-1];
+    }
+}
+
+///
+unittest
+{
+    auto gen = Random(unpredictableSeed);
+    auto rv = SimplexVariable!double(2);
+    double[3] x;
+    rv(gen, x);
+    assert(x[0] >= 0 && x[1] >= 0 && x[2] >= 0);
+    assert(fabs(x[0] + x[1] + x[2] - 1) < 1e-10);
+}
+
+/++
+$(Dirichlet distribution).
++/
+struct DirichletVariable(T, AlphaParams = const(T)[])
+    if (isFloatingPoint!T)
+{
+    ///
+    enum isRandomVariable = true;
+
+    private size_t dim;
+    private AlphaParams alpha;
+
+    /++
+    Params:
+        alpha = (array of) concentration parameters
+    Constraints: `alpha[i] > 0`
+    +/
+    this(AlphaParams alpha)
+    {
+        assert(alpha.length >= 1);
+        for(size_t i = 0; i < alpha.length; ++i)
+            assert(alpha[i] > T(0));
+
+        this.dim = alpha.length - 1;
+        this.alpha = alpha;
+    }
+
+    ///
+    void opCall(G)(ref G gen, T[] result)
+    {
+        assert(result.length == dim + 1);
+        opCall(gen, result.sliced);
+    }
+
+    ///
+    void opCall(G, SliceKind kind, Iterator)(ref G gen, Slice!(kind, [1], Iterator) result)
+        if (isSaturatedRandomEngine!G)
+    {
+        assert(result.length == dim + 1);
+        for(size_t i = 0; i < result.length; ++i)
+            result[i] = GammaVariable!T(alpha[i])(gen);
+        result[] /= result.sum!"kbn";
+    }
+}
+
+///
+unittest
+{
+    auto gen = Random(unpredictableSeed);
+    auto rv = DirichletVariable!double([1.0, 5.7, 0.3]);
+    double[3] x;
+    rv(gen, x);
+    assert(x[0] >= 0 && x[1] >= 0 && x[2] >= 0);
+    assert(fabs(x[0] + x[1] + x[2] - 1) < 1e-10);
+}
+
+/++
+ Compute Cholesky decomposition in place. Only accesses lower/left half of
+ the matrix. Returns false if the matrix is not positive definite.
+ +/
+private bool cholesky(SliceKind kind, Iterator)(Slice!(kind, [2], Iterator) m)
+    if(isFloatingPoint!(DeepElementType!(typeof(m))))
+{
+    alias dotm = reduce!"a - b * c"; // note the `-`
+    assert(m.length!0 == m.length!1);
+
+    /* this is a straight-forward implementation of the Cholesky-Crout algorithm
+    from https://en.wikipedia.org/wiki/Cholesky_decomposition#Computation */
+    for(size_t i = 0; i < m.length!0; ++i)
```

Comment:
something like that, to catche indexing operations:
```d
foreach(size_t; 0 .. m.length!0)
{
    auto mi = m[i];
    foreach(size_t j; 0 .. i)
    {
        auto mj = m[j]l
        mi[j] = (mi[j] - dot(mi[0 .. j], mj[0 .. j])) / mj[j];
    }
}
```
Also please add a private common `dot` functions and use it instead of all `reduce` in this module